beam/beam/demand/demand.py

```diff
@@ -134,64 +167,206 @@ def build_demand_map() -> None:
 	build_allocation_map()
 
 
-def build_allocation_map(doc=None, doc_row=None, action=None):
-	supply_allocation = []
-	demand_allocation = []
+def build_allocation_map(
+	row: Union[
+		"DeliveryNoteItem",
+		"PurchaseInvoiceItem",
+		"PurchaseReceiptItem",
+		"SalesInvoiceItem",
+		"StockEntryDetail",
+		"StockReconciliationItem",
+		None,
+	] = None,
+	action: dict | None = None,
+):
+	if row and action:
+		update_allocations(row=row, action=action)
+	else:
+		create_allocations()
+
+
+def get_demand_query(
+	cursor: "Cursor",
+	row: Union[
+		"DeliveryNoteItem",
+		"PurchaseInvoiceItem",
+		"PurchaseReceiptItem",
+		"SalesInvoiceItem",
+		"StockEntryDetail",
+		"StockReconciliationItem",
+		None,
+	] = None,
+	action: dict | None = None,
+):
+	item_query = warehouse_query = ""
+	if row:
+		item_query = f"WHERE item_code = '{row.item_code}'"
+		# TODO: should we only consider warehouse if it is the same as the demand warehouse?
+		# if action:
+		# 	warehouse_field = action.get("warehouse_field")
+		# 	if warehouse_field:
+		# 		warehouse_query = f"AND warehouse = '{row.get(warehouse_field)}'"
```

Comment:
@agritheory this part of the code was checking the transaction's warehouse against the demand warehouse, which I'm not sure if that's intentional? I've commented it out and that makes the allocations against non-demand warehouses to work correctly.

Reply:
Yeah, part of the demand feature is that it is only satisfied when it's in the correct warehouse. We can allocate before that, but the demand cannot be met unless the inventory is in the correct warehouse.

Reply:
So should I keep that code? Keeping it will only add inventory allocation in demand warehouses.

Reply:
No, allocations are warehouse-agnostic except for the excluded warehouses (quarantine /scrap/ subcontractor). We want to allocate from all warehouses except those. So I think that this code needs to be updated for that scenario but not a 1:1 matching one.